pkgs/expander/extract/gc-defn.rkt

```diff
@@ -158,35 +41,49 @@
      [(defn? e)
       (for ([sym (in-list (defn-syms e))])
         (hash-set! sym-to-rhs sym (defn-rhs e)))]))
-  
+
   ;; A "mark"-like traversal of an expression:
   (define (set-all-used! init-sym e)
     (for ([sym (in-set (all-used-symbols e))])
       (unless (hash-ref used-syms sym #f)
-        (when (hash-ref sym-to-rhs sym #f)
-          (log-info "adding ~s for because it's referenced by ~s" sym init-sym))
         (hash-set! used-syms sym #t)
         (set-all-used! sym (hash-ref sym-to-rhs sym #f)))))
-  
+
+  (define (ref? s) (hash-ref seen-defns s (lambda () #f)))
+
+  (define (safe-defn? e)
+    (and (defn? e)
+         (or (not (any-side-effects? (defn-rhs e) (length (defn-syms e)) ref?
+                                     #:known-defns seen-defns)))))
+
   ;; Mark each body form, delaying the righthand side of definitions
   ;; if the definition has no side-effect
-  (for ([e (in-list body)])
-    (cond
-     [(defn? e)
-      (begin0
-          (if (any-side-effects? (defn-rhs e) (length (defn-syms e)) (lambda (s) (hash-ref seen-defns s #f)))
-              (begin 
-                (log-info "adding syms for RHS side effects: ~s" (defn-syms e))
-                (set-all-used! (defn-syms e) (defn-rhs e)))
-              (seteq))
-        (for ([s (in-list (defn-syms e))])
-          (hash-set! seen-defns s #t)))]
-     [else (set-all-used! '<toplevel> e)]))
-  
+  (let loop ([body body])
+    (cond [(null? body) (void)]
+          [(defn? (car body))
+           (for* ([d (in-list body)]
+                    #:break (not (safe-defn? d))
+                    [s (in-list (defn-syms d))])
+             (unless (hash-ref seen-defns s #f)
+               (hash-set! seen-defns s #s(def))))
+           (define e (car body))
+           (define s
+             (cond [(any-side-effects? (defn-rhs e) (length (defn-syms e)) ref? #:known-defns seen-defns)
+                    (for ([sym (in-list (defn-syms e))])
```

Comment:
@mflatt This `for` loop is the bug fix I mentioned -- previously we weren't marking the actual definition as needing to stay.